File: app/src/main/java/com/ngengs/android/popularmovies/apps/data/local/MoviesProviderHelper.java
package com.ngengs.android.popularmovies.apps.data.local;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;

import com.ngengs.android.popularmovies.apps.data.MoviesDetail;
import com.ngengs.android.popularmovies.apps.data.MoviesList;
import com.ngengs.android.popularmovies.apps.globals.LocalDatabase;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by ngengs on 7/4/2017.
 */

@SuppressWarnings("DefaultFileTemplate")
public class MoviesProviderHelper {
    private final Context context;

    public MoviesProviderHelper(Context context) {
        this.context = context;
    }

    public MoviesList getFavorites() {
        Cursor cursor = context.getContentResolver().query(MoviesFavorites.CONTENT_URI, MoviesEntry.getColumnsWithTable(), null, null, null);
        List<MoviesDetail> data = new ArrayList<>();
        if (cursor != null) {
            while (cursor.moveToNext()) {
                data.add(MoviesDetail.fromCursor(cursor));
            }
            MoviesList moviesList = new MoviesList(data);
            moviesList.setPage(1);
            moviesList.setTotalPage(1);
            moviesList.setTotalResult(data.size());
            return moviesList;
        } else return null;
    }

    public MoviesList getPopular() {
        Cursor cursor = context.getContentResolver().query(MoviesPopular.CONTENT_URI, MoviesEntry.getColumnsWithTable(), null, null, null);
        List<MoviesDetail> data = new ArrayList<>();
        if (cursor != null) {
            while (cursor.moveToNext()) {
                data.add(MoviesDetail.fromCursor(cursor));
            }
            MoviesList moviesList = new MoviesList(data);
            moviesList.setPage(0);
            moviesList.setTotalPage(1);
            moviesList.setTotalResult(data.size());
            return moviesList;
        } else return null;
    }

    public void deletePopular() {
        context.getContentResolver().delete(MoviesPopular.CONTENT_URI, null, null);
    }

    public void savePopular(List<MoviesDetail> movies) {
        if (!movies.isEmpty()) {
            ContentValues[] contentValues = new ContentValues[movies.size()];
            for (int i = 0; i < movies.size(); i++) {
                ContentValues contentValue = new ContentValues();
                contentValue.put(LocalDatabase.COLUMN_MOVIE_ID_KEY, movies.get(i).getId());
                contentValues[i] = contentValue;
            }
            context.getContentResolver().bulkInsert(MoviesPopular.CONTENT_URI, contentValues);
        }
    }

    public MoviesList getTopRated() {
        Cursor cursor = context.getContentResolver().query(MoviesTopRated.CONTENT_URI, MoviesEntry.getColumnsWithTable(), null, null, null);
        List<MoviesDetail> data = new ArrayList<>();
        if (cursor != null) {
            while (cursor.moveToNext()) {
                data.add(MoviesDetail.fromCursor(cursor));
            }
            MoviesList moviesList = new MoviesList(data);
            moviesList.setPage(0);
            moviesList.setTotalPage(1);
            moviesList.setTotalResult(data.size());
            return moviesList;
        } else return null;
    }

    public void deleteTopRated() {
        context.getContentResolver().delete(MoviesTopRated.CONTENT_URI, null, null);
    }

    public void saveTopRated(List<MoviesDetail> movies) {
        if (!movies.isEmpty()) {
            ContentValues[] contentValues = new ContentValues[movies.size()];
            for (int i = 0; i < movies.size(); i++) {
                ContentValues contentValue = new ContentValues();
                contentValue.put(LocalDatabase.COLUMN_MOVIE_ID_KEY, movies.get(i).getId());
                contentValues[i] = contentValue;
            }
            context.getContentResolver().bulkInsert(MoviesTopRated.CONTENT_URI, contentValues);
        }
    }

    public void saveMovies(List<MoviesDetail> movies) {
        if (!movies.isEmpty()) {
            ContentValues[] contentValues = new ContentValues[movies.size()];
            for (int i = 0; i < movies.size(); i++) {
                contentValues[i] = movies.get(i).toContentValues();
            }
            context.getContentResolver().bulkInsert(MoviesEntry.CONTENT_URI, contentValues);
        }
    }

    public void saveMovies(MoviesDetail movie) {
        context.getContentResolver().insert(MoviesEntry.CONTENT_URI, movie.toContentValues());
    }

    public void addToFavorites(int movieId) {
        ContentValues contentValues = new ContentValues();
        contentValues.put(LocalDatabase.COLUMN_MOVIE_ID_KEY, movieId);
        context.getContentResolver().insert(MoviesFavorites.CONTENT_URI, contentValues);
    }

    public void removeFromFavorites(int movieId) {
        context.getContentResolver().delete(
                MoviesFavorites.CONTENT_URI,
                LocalDatabase.COLUMN_MOVIE_ID_KEY + " = " + movieId,
                null
        );
    }

    public boolean isFavorite(int movieId) {
        boolean favorite = false;
        Cursor cursor = context.getContentResolver().query(
                MoviesFavorites.CONTENT_URI,
                null,
                LocalDatabase.COLUMN_MOVIE_ID_KEY + " = " + movieId,
                null,
                null
        );
        if (cursor != null) {
            favorite = cursor.getCount() != 0;
            cursor.close();
        }
        return favorite;
    }
}

File: app/src/main/java/com/ngengs/android/popularmovies/apps/data/local/MoviesProviderHelper.kt
package com.ngengs.android.popularmovies.apps.data.local

import android.content.ContentValues
import android.content.Context
import com.ngengs.android.popularmovies.apps.data.remote.MoviesDetail
import com.ngengs.android.popularmovies.apps.data.remote.MoviesList
import com.ngengs.android.popularmovies.apps.globals.LocalDatabase

/**
 * Created by rizky.kharisma on 10/11/22.
 * @ngengs
 */
class MoviesProviderHelper(private val context: Context) {
    fun getFavorites(): MoviesList? {
        val cursor = context.contentResolver.query(
            MoviesFavorites.CONTENT_URI,
            MoviesEntry.getColumnsWithTable(),
            null,
            null,
            null
        )
        val data: MutableList<MoviesDetail> = mutableListOf()
        return if (cursor != null) {
            while (cursor.moveToNext()) {
                data.add(MoviesDetail.fromCursor(cursor))
            }
            MoviesList(1, data.size, 1, data)
        } else null
    }

    fun getPopular(): MoviesList? {
        val cursor = context.contentResolver.query(
            MoviesPopular.CONTENT_URI,
            MoviesEntry.getColumnsWithTable(),
            null,
            null,
            null
        )
        val data: MutableList<MoviesDetail> = mutableListOf()
        return if (cursor != null) {
            while (cursor.moveToNext()) {
                data.add(MoviesDetail.fromCursor(cursor))
            }
            MoviesList(0, data.size, 1, data)
        } else null
    }

    fun deletePopular() {
        context.contentResolver.delete(MoviesPopular.CONTENT_URI, null, null)
    }

    fun savePopular(movies: List<MoviesDetail>) {
        if (movies.isNotEmpty()) {
            val contentValues = arrayOfNulls<ContentValues>(movies.size)
            for (i in movies.indices) {
                val contentValue = ContentValues()
                contentValue.put(LocalDatabase.COLUMN_MOVIE_ID_KEY, movies[i].id)
                contentValues[i] = contentValue
            }
            context.contentResolver.bulkInsert(MoviesPopular.CONTENT_URI, contentValues)
        }
    }

    fun getTopRated(): MoviesList? {
        val cursor = context.contentResolver.query(
            MoviesTopRated.CONTENT_URI,
            MoviesEntry.getColumnsWithTable(),
            null,
            null,
            null
        )
        val data: MutableList<MoviesDetail> = mutableListOf()
        return if (cursor != null) {
            while (cursor.moveToNext()) {
                data.add(MoviesDetail.fromCursor(cursor))
            }
            MoviesList(0, data.size, 1, data)
        } else null
    }

    fun deleteTopRated() {
        context.contentResolver.delete(MoviesTopRated.CONTENT_URI, null, null)
    }

    fun saveTopRated(movies: List<MoviesDetail>) {
        if (movies.isNotEmpty()) {
            val contentValues = arrayOfNulls<ContentValues>(movies.size)
            for (i in movies.indices) {
                val contentValue = ContentValues()
                contentValue.put(LocalDatabase.COLUMN_MOVIE_ID_KEY, movies[i].id)
                contentValues[i] = contentValue
            }
            context.contentResolver.bulkInsert(MoviesTopRated.CONTENT_URI, contentValues)
        }
    }

    fun saveMovies(movies: List<MoviesDetail>) {
        if (movies.isNotEmpty()) {
            val contentValues = arrayOfNulls<ContentValues>(movies.size)
            for (i in movies.indices) {
                contentValues[i] = movies[i].toContentValues()
            }
            context.contentResolver.bulkInsert(MoviesEntry.CONTENT_URI, contentValues)
        }
    }

    fun saveMovies(movie: MoviesDetail) {
        context.contentResolver.insert(MoviesEntry.CONTENT_URI, movie.toContentValues())
    }

    fun addToFavorites(movieId: Int) {
        val contentValues = ContentValues()
        contentValues.put(LocalDatabase.COLUMN_MOVIE_ID_KEY, movieId)
        context.contentResolver.insert(MoviesFavorites.CONTENT_URI, contentValues)
    }

    fun removeFromFavorites(movieId: Int) {
        context.contentResolver.delete(
            MoviesFavorites.CONTENT_URI,
            LocalDatabase.COLUMN_MOVIE_ID_KEY + " = " + movieId,
            null
        )
    }

    fun isFavorite(movieId: Int): Boolean {
        var favorite = false
        val cursor = context.contentResolver.query(
            MoviesFavorites.CONTENT_URI,
            null,
            LocalDatabase.COLUMN_MOVIE_ID_KEY + " = " + movieId,
            null,
            null
        )
        if (cursor != null) {
            favorite = cursor.count != 0
            cursor.close()
        }
        return favorite
    }
}